File: Java/프로그래머스/Solution_pg_메뉴리뉴얼.java
package programmers;

import java.util.*;
import java.util.Map.Entry;

public class Solution_pg_메뉴리뉴얼 {

	public static void main(String[] args) {
		String[] orders = {"XYZ","XWY","WXA"};
		int[] course = {2,3,4};
		String[] answer = solution(orders, course);
		
		print(answer);
	}
	static int N, R, max, idx;
	static char[] menu, pick;
	static HashMap<String, Integer> map;
	public static String[] solution(String[] orders, int[] course) {
		int len = orders.length;
		for(int i=0;i<len;i++) {
			char[] arr = orders[i].toCharArray();
			Arrays.sort(arr);
			StringBuilder sb = new StringBuilder();
			for(char c : arr)
				sb.append(c);
			orders[i] = sb.toString();
		}
		
		ArrayList<String> list = new ArrayList<String>();
		for(int n : course) {
			R = n;
			max = 0;
			pick = new char[R];
			map = new HashMap<String, Integer>();
			for(String order: orders) {
				menu = order.toCharArray();
				N = order.length();
				combination(0,0);
			}
			idx++;
			
			if(max > 1) {
				Set<Entry<String, Integer>> set = map.entrySet();
				Iterator<Entry<String, Integer>> iter = set.iterator();
				while(iter.hasNext()) {
					Entry<String, Integer> entry = iter.next();
					if(entry.getValue() == max)
						list.add(entry.getKey());
				}
			}
		}
		
		len = list.size();
		String[] answer = new String[len];
		for(int i=0;i<len;i++)
			answer[i] = list.get(i);
		Arrays.sort(answer);
		
		return answer;
    }
	private static void combination(int cnt, int start) {
		if(cnt == R) {
			StringBuilder sb = new StringBuilder();
			for(char c : pick)
				sb.append(c);
			String s = sb.toString();
			if(map.containsKey(s))
				map.put(s, map.get(s)+1);
			else
				map.put(s, 1);
			
			max = Math.max(max, map.get(s));
			return;
		}
		for(int i=start;i<N;i++) {
			pick[cnt] = menu[i];
			combination(cnt+1, i+1);
		}
	}
	private static void print(String[] arr) {
		for(String s : arr)
			System.out.print(s+" ");
		System.out.println();
	}
}

File: JS/programmers/solution_pg_메뉴리뉴얼.js
let N, R, order, menu, map, max;
function solution(orders, course) {
  for (let i in orders) {
    let arr = orders[i].split('');
    arr.sort();
    orders[i] = '';
    for (let a of arr) orders[i] += a;
  }
  let answer = [];
  for (let c of course) {
    R = c;
    max = 0;
    map = new Map();
    for (let o of orders) {
      order = o;
      N = order.length;
      menu = new Array(R);
      combination(0, 0);
    }
    if (max > 1) {
      for (let [key, value] of map) {
        if (value == max) answer.push(key);
      }
    }
  }
  answer.sort();
  return answer;
}
function combination(start, count) {
  if (count == R) {
    let newMenu = '';
    for (let i = 0; i < R; i++) newMenu += menu[i];
    if (map.has(newMenu)) map.set(newMenu, map.get(newMenu) + 1);
    else map.set(newMenu, 1);
    max = Math.max(max, map.get(newMenu));
    return;
  }
  for (let i = start; i < N; i++) {
    menu[count] = order.charAt(i);
    combination(i + 1, count + 1);
  }
}
